apollo-router-core/src/query_cache.rs

```diff
@@ -0,0 +1,107 @@
+use crate::prelude::graphql::*;
+use futures::lock::Mutex;
+use lru::LruCache;
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::sync::broadcast;
+
+/// A cache for parsed GraphQL queries.
+#[derive(Debug)]
+pub struct QueryCache {
```

Comment:
considering how tricky that code can be, we should refactor that in a separate cache type

Reply:
to have one canonical cache implementation

Reply:
Yes I agree. As I said I plan to do this in a later PR. Unless you think there is an argument for doing this here I will consider this out-of-scope for now.

Reply:
yep, it can be done in another PR

Reply:
It's also worth checking that a loading cache implementation does not exist already. This one looks promising, although I can't vouch for the internals.

https://github.com/zerotwo-bot/cache-loader-async-rs